File: BinaryTree/ts/leafList.ts
import { TreeNode } from "./TreeNode";

export function leafList(root: TreeNode<number> | null): number[] {
  const leaves: number[] = [];
  collectLeaves(root, leaves);
  return leaves;
}

function collectLeaves(node: TreeNode<number> | null, leaves: number[]): void {
  if (node === null) {
    return;
  }
  if (node.left === null && node.right === null) {
    leaves.push(node.val);
    return;
  }
  collectLeaves(node.left, leaves);
  collectLeaves(node.right, leaves);
}

File: BinaryTree/go/leafList.go
package binarytree

// LeafList returns the values of every leaf from left to right.
func LeafList(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	var leaves []int
	collectLeaves(root, &leaves)
	return leaves
}

func collectLeaves(node *TreeNode, leaves *[]int) {
	if node == nil {
		return
	}
	if node.Left == nil && node.Right == nil {
		*leaves = append(*leaves, node.Val)
		return
	}
	collectLeaves(node.Left, leaves)
	collectLeaves(node.Right, leaves)
}
